language/vm/src/deserializer.rs

```diff
@@ -1510,3 +1517,219 @@ impl Opcodes {
+                    SignatureToken::Signer
+                }
+                // Map signer to signer ref to keep invalid scripts invalid
+                SignatureToken::Signer => {
+                    SignatureToken::Reference(Box::new(SignatureToken::Signer))
+                }
+                _ => s.clone(),
+            })
+            .collect::<Vec<_>>(),
+    );
+
+    // Make a new local signature
+    let (new_locals_signature, new_local_idxs) = {
+        let mut new_sig = (0..signer_ref_parameters.len())
+            .map(|_| SignatureToken::Reference(Box::new(SignatureToken::Signer)))
+            .collect::<Vec<_>>();
+        new_sig.extend(locals_signature.0.iter().cloned());
+        let new_local_idxs = (0..signer_ref_parameters.len())
+            .map(|idx| idx as LocalIndex + num_parameters)
+            .collect::<Vec<_>>();
+        (Signature(new_sig), new_local_idxs)
+    };
+
+    // Determine a mapping from the owned signers to the new locals with the borrowed signer param
+    // Build instructions to borrow those params
+    let (signer_ref_remapping, mut borrow_owned_signers_instructions) = {
+        let mut new_code = vec![];
+        let mut signer_ref_remapping = (0..num_parameters).map(|_| None).collect::<Vec<_>>();
+
+        for (signer_ref_parameter_idx, new_local_idx) in
+            signer_ref_parameters.iter().zip(&new_local_idxs)
+        {
+            signer_ref_remapping[*signer_ref_parameter_idx as usize] = Some(new_local_idx);
+            new_code.push(Bytecode::ImmBorrowLoc(*signer_ref_parameter_idx));
+            new_code.push(Bytecode::StLoc(*new_local_idx));
+        }
+        (signer_ref_remapping, new_code)
+    };
+
+    // Update the code
+    let number_of_new_locals = signer_ref_parameters.len() as LocalIndex;
+    let number_of_new_instructions = borrow_owned_signers_instructions.len() as CodeOffset;
+    for instr in &mut script.code.code {
+        match instr {
+            // increment the offsets based on the number of new instructions added
+            Bytecode::BrTrue(offset) | Bytecode::BrFalse(offset) | Bytecode::Branch(offset) => {
+                *offset += number_of_new_instructions
+            }
+
+            // If the local used was an old signer reference, remapp it to the new local borrowing
+            // the owned signer
+            Bytecode::CopyLoc(local_idx)
+            | Bytecode::MoveLoc(local_idx)
+            | Bytecode::StLoc(local_idx)
+            | Bytecode::MutBorrowLoc(local_idx)
+            | Bytecode::ImmBorrowLoc(local_idx) => {
+                if *local_idx < num_parameters {
+                    if let Some(new_local) = signer_ref_remapping[*local_idx as usize] {
+                        *local_idx = *new_local
+                    }
+                } else {
+                    *local_idx += number_of_new_locals
+                }
+            }
+            Bytecode::Pop
```

Comment:
I'm normally the world's biggest fan of exhaustive pattern matching, but this might be one case where it makes more sense not to have it... Instructions with a `local_idx` will never appear in `VERSION_1` code, and everyone who adds a new instruction will have to add an unused case here.